retriever/lib/store.js

```diff
@@ -112,17 +124,35 @@ export async function getOwnerByRootCid(env, rootCid) {
       error: `Root_cid '${rootCid}' exists but has no associated owner.`,
     }
   }
-  const approved = withOwner.find((row) => approvedOwners.includes(row.owner))
-  if (!approved) {
+  const approved = withOwner.filter((row) => approvedOwners.includes(row.owner))
+  if (approved.length === 0) {
     return {
       error: `Root_cid '${rootCid}' exists but has no approved owner`,
     }
   }
+  const withPaymentRail = approved.filter(
+    (row) => row.payer && row.payer.toLowerCase() === clientAddress,
+  )
+  if (withPaymentRail.length === 0) {
+    return {
+      error: `There is no payment rail for client '${clientAddress}' and root_cid '${rootCid}'.`,
+    }
+  }
+
+  const withCDN = withPaymentRail.filter(
+    (row) => row.with_cdn && row.with_cdn === 1,
+  )
+  if (withCDN.length === 0) {
+    return {
+      error: `The payment rail for client '${clientAddress}' and root_cid '${rootCid}' has withCDN=false.`,
+      errorStatus: 402,
+    }
+  }
 
-  const { set_id: setId, owner } = approved
+  const { set_id: setId, owner } = withCDN[0]
```

Comment:
We could also use destructuring on the array instead of index (just seems a bit more idiomatic to me).